patterns/multiparty-note-sharing/broadcaster/src/main.nr

```diff
@@ -0,0 +1,180 @@
+    2) nullify_note(): which will nullify the note. In the tests we will show all accounts with the note can nullify the note.
+    3) broadcast_note_for(): will take recipients, an id to identify a note, and will broadcast it to the recipients.
+
+    Please note: 
+    - The asynchronous benefit of this pattern is not currently supported.
+    - The code shown here is a toy example made with the objective of fully isolating the pattern from any specific business logic.
+*/
+
+    // Declaring the slots for the notes
+    global SHARED_NOTE_SLOT = 1;
+
+    // Declaring storage
+    struct Storage {
+        shared_notes: Set<SharedNote, SHARED_NOTE_LEN>,
+    }
+
+    // Declaring storage implementation
+    impl Storage {
+        fn init(context: Context) -> Self {
+            Storage {
+                shared_notes: Set::new(context, SHARED_NOTE_SLOT, SharedNoteMethods),
+            }
+        }
+    }
+ 
+    /// @notice Constructor for the contract.
+    #[aztec(private)]
+    fn constructor() {}
+
+    /// @notice Creates a SharedNote.
+    #[aztec(private)]
+    fn create_note() -> Field {
+        // Store the sender for readability.
+        let sender = context.msg_sender();
+
+        // Create a random shared nullifier key
+        let shared_nullifier = rand();
+
+        // Create the note.
+        let mut shared_note = SharedNote::new(sender, shared_nullifier);
+
+        // Insert it into storage and broadcast it to the owner.
+        storage.shared_notes.insert(&mut shared_note, true);
+
+        // Return the shared nullifier
+        shared_nullifier
+    }
+
+    /// @notice Consumes a SharedNote.
+    /// @param shared_nullifier The shared_nullifier used when creating the note.
+    /// @param nonce The authwit nonce.
+    #[aztec(private)]
+    fn nullify_note(shared_nullifier: Field, nonce: Field) {
+        // Set the sender for readability.
+        let sender = context.msg_sender();
+
+        // Create a filter to fetch the note.
+        let shared_note_filter = NoteGetterOptions::new().select(1, shared_nullifier).set_limit(1);
+
+        // Get the note.
+        let mut shared_note = storage.shared_notes.get_notes(shared_note_filter)[0];
+
+        // Ensure the note exists.
+        assert(!shared_note.is_none(), "Note doesnt exist");
+
+        let shared_note = shared_note.unwrap_unchecked();
+
+        // If the sender is not the owner, ensure the caller is authorized to call this contract.
+        // For users that have notes where the owner is a contract, this forces to go through the contract to nullify the note.
+        if(!sender.eq(shared_note.owner)) {
+            assert_current_call_valid_authwit(&mut context, shared_note.owner);
+        } else {
+            assert(nonce == 0, "Invalid nonce");
+        }
+
+        // This extra constraint is needed to avoid a compiler bug
```

Comment:
That's maybe a bit too opiniated? Is it working if we include the previous check in the if block? Maybe tidier to then have 
```
if owner != 0:
  authwit?
  etc
else "does not exist"
```

Reply:
Changed `bug` for `error` and added a bit of context, but left the checks separated